JuliaStream.jl/src/Stream.jl

```diff
@@ -0,0 +1,292 @@
+using ArgParse
+using Parameters
+using Printf
+using Base: Float64, Int
+
+include("StreamData.jl")
+
+const VectorData = StreamData{T,Vector{T}} where {T}
+
+struct Timings
+  copy::Vector{Float64}
+  mul::Vector{Float64}
+  add::Vector{Float64}
+  triad::Vector{Float64}
+  dot::Vector{Float64}
+  Timings(n) = new(zeros(n), zeros(n), zeros(n), zeros(n), zeros(n))
+end
+
+@enum Benchmark All Triad Nstream
+
+function run_all!(data::StreamData{T,C}, times::Int)::Tuple{Timings,T} where {T,C}
+  timings = Timings(times)
+  lastSum::T = 0
+  for i = 1:times
+    @inbounds timings.copy[i] = @elapsed copy!(data)
+    @inbounds timings.mul[i] = @elapsed mul!(data)
+    @inbounds timings.add[i] = @elapsed add!(data)
+    @inbounds timings.triad[i] = @elapsed triad!(data)
+    @inbounds timings.dot[i] = @elapsed lastSum = dot(data)
+  end
+  return (timings, lastSum)
+end
+
+function run_triad!(data::StreamData{T,C}, times::Int)::Float64 where {T,C}
+  return @elapsed for _ = 1:times
+    triad!(data)
+  end
+end
+
+function run_nstream!(data::StreamData{T,C}, times::Int)::Vector{Float64} where {T,C}
+  timings::Vector{Float64} = zeros(times)
+  for i = 1:times
+    @inbounds timings[i] = @elapsed nstream!(data)
+  end
+  return timings
+end
+
+function check_solutions(
+  data::StreamData{T,C},
+  times::Int,
+  init::Tuple{T,T,T},
+  benchmark::Benchmark,
+  dot::Union{T,Nothing},
+) where {T,C}
+  (gold_a, gold_b, gold_c) = init
+  for _ = 1:times
+    if benchmark == All
+      gold_c = gold_a
+      gold_b = data.scalar * gold_c
+      gold_c = gold_a + gold_b
+      gold_a = gold_b + data.scalar * gold_c
+    elseif benchmark == Triad
+      gold_a = gold_b + data.scalar * gold_c
+    elseif benchmark == Nstream
+      gold_a += gold_b + data.scalar * gold_c
+    else
+      error("Unknown benchmark", benchmark)
+    end
+  end
+
+  tolerance = eps(T) * 100
```

Comment:
Do we need to worry about making sure `100` is floating point here?

Reply:
I haven't used Julia long enough to know whether doing it like this (via multiple dispatch of the `*` function) is idiomatic. Since Julia's documentation mentions multiple dispatch as Julia's forte, I think this is probably an OK way.
It's also safe:
```
julia> typeof(eps(Float64) * 1)
Float64
``` 
The alternative would be to `convert` first:
```
julia> typeof(convert(Float64, 100))
Float64
julia> eps(Float64) * convert(Float64, 100)
2.220446049250313e-14
```
I'm leaning towards keeping `* 100`